index.js

```diff
@@ -28,10 +28,36 @@ function apply(options, compiler) {
 
   var runner = runCompilation.bind(this, options);
 
-  compiler.plugin('run', runner);
-  compiler.plugin('watch-run', function onWatchRun(watcher, callback) {
-    runner(watcher.compiler, callback);
-  });
+  if (options.lintDirtyModulesOnly) {
+    this.startTime = Date.now();
+    this.prevTimestamps = {};
+    var isFirstRun = true;
+
+    compiler.plugin('emit', function (compilation, callback) {
+      var dirtyOptions = assign({}, options);
+      var globPatterb = dirtyOptions.files;
+
+      var changedFiles = Object.keys(compilation.fileTimestamps).filter(function (watchfile) {
+        return (this.prevTimestamps[watchfile] || this.startTime) < (compilation.fileTimestamps[watchfile] || Infinity);
+      }.bind(this)).filter(minimatch.filter(globPatterb.join('|'), {matchBase: true}));
```

Comment:
You could even bring in `lodash.reduce` and split out this function into a private module in `lib`, since it's quite complex, and probably will have some compatibility issues with webpack 2 (#46).